server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/message/view/MessageFastViewFactory.java

```diff
@@ -0,0 +1,169 @@
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+public class MessageFastViewFactory implements MessageViewFactory<MessageFastView> {
+
+    static class FromMessageResultAndPreview implements Helpers.FromMessageResult<MessageFastView> {
+
+        private final BlobManager blobManager;
+        private final MessageFullViewFactory messageFullViewFactory;
+        private final Map<MessageId, MessageFastViewPrecomputedProperties> fastProjections;
+
+        FromMessageResultAndPreview(BlobManager blobManager, MessageFullViewFactory messageFullViewFactory,
+                                    Map<MessageId, MessageFastViewPrecomputedProperties> fastProjections) {
+            this.blobManager = blobManager;
+            this.messageFullViewFactory = messageFullViewFactory;
+            this.fastProjections = fastProjections;
+        }
+
+        @Override
+        public MessageFastView fromMessageResults(Collection<MessageResult> messageResults) throws MailboxException, IOException {
+            Helpers.assertOneMessageId(messageResults);
+
+            MessageResult firstMessageResult = messageResults.iterator().next();
+            List<MailboxId> mailboxIds = Helpers.getMailboxIds(messageResults);
+
+            Message mimeMessage = Helpers.parse(firstMessageResult.getFullContent().getInputStream());
+
+            return MessageFastView.builder()
+                .id(firstMessageResult.getMessageId())
+                .mailboxIds(mailboxIds)
+                .blobId(BlobId.of(blobManager.toBlobId(firstMessageResult.getMessageId())))
+                .threadId(firstMessageResult.getMessageId().serialize())
+                .keywords(Helpers.getKeywords(messageResults))
+                .size(firstMessageResult.getSize())
+                .inReplyToMessageId(Helpers.getHeaderValue(mimeMessage, "in-reply-to"))
+                .subject(Strings.nullToEmpty(mimeMessage.getSubject()).trim())
+                .headers(Helpers.toHeaderMap(mimeMessage.getHeader().getFields()))
+                .from(Emailer.firstFromMailboxList(mimeMessage.getFrom()))
+                .to(Emailer.fromAddressList(mimeMessage.getTo()))
+                .cc(Emailer.fromAddressList(mimeMessage.getCc()))
+                .bcc(Emailer.fromAddressList(mimeMessage.getBcc()))
+                .replyTo(Emailer.fromAddressList(mimeMessage.getReplyTo()))
+                .date(Helpers.getDateFromHeaderOrInternalDateOtherwise(mimeMessage, firstMessageResult))
+                .preview(getPreview(firstMessageResult.getMessageId(), mimeMessage))
+                .build();
+        }
+
+        private Optional<Preview> getPreview(MessageId messageId, Message mimeMessage) {
+            return OptionalUtils.orSuppliers(
+                () -> Optional.ofNullable(fastProjections.get(messageId))
+                    .map(MessageFastViewPrecomputedProperties::getPreview),
+                () -> computePreview(mimeMessage));
```

Comment:
No. We don't have the body here!

We thus cannot compute the preview.

Please add a precondition in `MessageFastViewFactory.FromMessageResultAndPreview::fromMessageResults` to ensure this.fastProjections contains an entry for that messageId.

That code path should never be used by the way!

Reply:
That's **temporal coupling**!

The MimeMessage will have a Body allowing to compute the preview only if the message result was read with FetchGroup.Full. Here *you don't know if the MessageResult is read with the right fetchGroup*, as it's actually called with HEADERS and FULL FetchGroups. And I thind that to be especially tricky to reason about. I'm expecting bugs!

Reply:
That's **temporal coupling**!

The MimeMessage will have a Body allowing to compute the preview only if the message result was read with FetchGroup.Full. Here *you don't know if the MessageResult is read with the right fetchGroup*, as it's actually called with HEADERS and FULL FetchGroups. And I thind that to be especially tricky to reason about. I'm expecting bugs!